### PR TITLE
[Repo] Added initial TracerProviderBuilder dependency injection docs

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

Added initial `TracerProviderBuilder` dependency injection docs

### CODE REVIEW

File: docs/trace/customizing-the-sdk/README.md [hunk 1]
```diff
@@ -359,10 +359,143 @@ using OpenTelemetry;
...
+
+### Overview
+
+**Note** This information applies to the OpenTelemetry SDK version 1.4.0 and
+newer only.
+
+The SDK implementation of `TracerProviderBuilder` is backed by an
+`IServiceCollection` and supports a wide range of APIs to enable what is
+generally known as [dependency
+injection](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection).
```
Comment (cijothomas) [inline docs/trace/customizing-the-sdk/README.md:372]:
```suggestion
injection](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection).
```

File: docs/trace/customizing-the-sdk/README.md [hunk 2]
```diff
@@ -359,10 +359,143 @@ using OpenTelemetry;
...
+        {
+            // Implementation not important
+        }
+    }
+```
+
+We want to inject the `ILogger<MyCustomExporter>` and `MyCustomService`
+dependencies into our `MyCustomExporter` instance.
+
+#### Using Sdk.CreateTracerProviderBuilder()
```
Comment (cijothomas) [inline docs/trace/customizing-the-sdk/README.md:391]:
```suggestion
### Using Sdk.CreateTracerProviderBuilder()
```

File: docs/trace/customizing-the-sdk/README.md [hunk 3]
```diff
@@ -359,10 +359,143 @@ using OpenTelemetry;
...
+registered into that collection.
+
+#### Using the [OpenTelemetry.Extensions.Hosting](../../../src/OpenTelemetry.Extensions.Hosting/README.md) package
+
+**Note** If you are authoring an ASP.NET Core application or using the [.NET
+Generic Host](https://learn.microsoft.com/dotnet/core/extensions/generic-host)
+this is the recommended mechanism.
+
+```csharp
+builder.Services.AddSingleton<MyCustomService>();
```
Comment (cijothomas) [inline docs/trace/customizing-the-sdk/README.md:418]: possible confusion with this build and the otel builder.. lets use a different name?

File: docs/trace/customizing-the-sdk/README.md [hunk 4]
```diff
@@ -359,10 +359,143 @@ using OpenTelemetry;
...
+It will be able to access all services registered into that collection.
+Typically the collection used is the one managed by the application host.
+
+**Note** Multiple calls to `AddOpenTelemetryTracing` will configure the same
+`TracerProvider`. Only a single `TraceProvider` may exist in an
+`IServiceCollection` \ `IServiceProvider`.
+
+#### Using an external `IServiceCollection`
+
+To configure tracing in an external `IServiceCollection` use the
```
Comment (cijothomas) [inline docs/trace/customizing-the-sdk/README.md:435]: not sure i follow this part.. can you just merge this with the just-above section?

File: docs/trace/customizing-the-sdk/README.md (hunk 1 above)
Comment (CodeBlanch) [reply docs/trace/customizing-the-sdk/README.md]: Updated

File: docs/trace/customizing-the-sdk/README.md (hunk 2 above)
Comment (CodeBlanch) [reply docs/trace/customizing-the-sdk/README.md]: Updated

File: docs/trace/customizing-the-sdk/README.md (hunk 3 above)
Comment (CodeBlanch) [reply docs/trace/customizing-the-sdk/README.md]: Updated

File: docs/trace/customizing-the-sdk/README.md (hunk 4 above)
Comment (CodeBlanch) [reply docs/trace/customizing-the-sdk/README.md]: I removed it for now. I'll mention `ConfigureOpenTelemetryTracing` in the `Guidance for library authors` section in my next PR.

File: docs/trace/customizing-the-sdk/README.md [hunk 5]
```diff
@@ -359,10 +359,135 @@ using OpenTelemetry;
...
+](https://learn.microsoft.com/dotnet/api/microsoft.extensions.hosting.ihostedservice)
+is used to automatically start the `TracerProvider` when the host starts and the
+host will automatically shutdown and dispose the `TracerProvider` when it is
+shutdown.
+
+**Note** Multiple calls to `AddOpenTelemetryTracing` will configure the same
+`TracerProvider`. Only a single `TraceProvider` may exist in an
+`IServiceCollection` \ `IServiceProvider`.
+
+### Dependency injection `TracerProviderBuilder` extension methods
```
Comment (utpilla) [inline docs/trace/customizing-the-sdk/README.md:446]:
```suggestion
### Dependency injection using `TracerProviderBuilder` extension methods
```

File: docs/trace/customizing-the-sdk/README.md [hunk 6]
```diff
@@ -359,10 +359,135 @@ using OpenTelemetry;
...
+
+* `AddProcessor<T>`: Adds a processor of type `T` (must derive from
+  `BaseProcessor<Activity>`) into the `TracerProvider`.
+
+* `ConfigureServices`: Registers a callback function for configuring the
+  `IServiceCollection` used by the `TracerProviderBuilder`. **Note**
+  `ConfigureServices` may only be called before the `IServiceProvider` has been
+  created after which point service can no longer be added.
+
+* `ConfigureBuilder`: Registers a callback function for configuring the
```
Comment (utpilla) [inline docs/trace/customizing-the-sdk/README.md:462]: Maybe move this to the end as it's not expected to be used for regular usage?

[the remark this answers is not in the document]
Comment (CodeBlanch) [reply docs/trace/customizing-the-sdk/README.md:462]: It used to be the case, but not any longer! All the stuff here just needs SDK 😄

Comment (CodeBlanch) [reply docs/trace/customizing-the-sdk/README.md:467]: I moved `SetSampler` up so `ConfigureBuilder` is now the last thing. I agree with this fully but I suspect someone will come along and reorder alphabetically again 😄

File: docs/trace/customizing-the-sdk/README.md (hunk 5 above)
Comment (CodeBlanch) [reply docs/trace/customizing-the-sdk/README.md]: I don't really want this section to mix in with the "using" sections. I added a new heading to hopefully make it more clear. It is now like this...

```
### Overview
### Examples
#### Using Sdk.CreateTracerProviderBuilder()
#### Using the OpenTelemetry.Extensions.Hosting package
### Dependency injection `TracerProviderBuilder` extension method reference
### Guidance for library authors
```

How does that look?

**Review by utpilla — APPROVED** [review]

**Pull request merged**